File: plugin/src/main/java/io/github/kdy05/physicalFighters/ability/impl/CP9.java
package io.github.kdy05.physicalFighters.ability.impl;

import io.github.kdy05.physicalFighters.ability.Ability;
import io.github.kdy05.physicalFighters.ability.AbilitySpec;
import io.github.kdy05.physicalFighters.game.EventManager;
import io.github.kdy05.physicalFighters.game.InvincibilityManager;
import io.github.kdy05.physicalFighters.ability.AbilityUtils;
import io.github.kdy05.physicalFighters.util.EventData;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.util.Vector;

public class CP9 extends Ability {
    public CP9() {
        super(AbilitySpec.builder("CP9", Type.Active_Immediately, Rank.SS)
                .guide(Usage.IronAttack + "상대에게 6의 고정 대미지를 줍니다.",
                        Usage.IronRight + "바라보는 방향으로 빠르게 도약합니다.",
                        Usage.Passive + "낙하 대미지를 무시합니다.")
                .cooldown(20)
                .build());
        EventManager.onEntityDamageByEntity.add(new EventData(this, 0));
        registerRightClickEvent();
        EventManager.onEntityDamage.add(new EventData(this, 2));
    }

    @Override
    public int A_Condition(Event event, int CustomData) {
        if (CustomData == 0) {
            EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
            if (event0.getEntity() instanceof LivingEntity && isOwner(event0.getDamager())
                    && isValidItem(Ability.DefaultItem) && !InvincibilityManager.isDamageGuard()) {
                return 0;
            }
        } else if (CustomData == 1) {
            PlayerInteractEvent event1 = (PlayerInteractEvent) event;
            if (isOwner(event1.getPlayer()) && isValidItem(Ability.DefaultItem)) {
                return 1;
            }
        } else if (CustomData == 2) {
            EntityDamageEvent event2 = (EntityDamageEvent) event;
            if (isOwner(event2.getEntity()) && event2.getCause() == DamageCause.FALL) {
                event2.setCancelled(true);
                sendMessage(ChatColor.GREEN + "사뿐하게 떨어져 대미지를 받지 않았습니다.");
            }
        }
        return -1;
    }

    @Override
    public void A_Effect(Event event, int CustomData) {
        if (CustomData == 0) {
            EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
            LivingEntity entity = (LivingEntity) event0.getEntity();
            AbilityUtils.piercingDamage(entity, 6);
            sendMessage(String.format(ChatColor.RED + "%s에게 지건을 사용했습니다.", entity.getName()));
        } else if (CustomData == 1) {
            PlayerInteractEvent event1 = (PlayerInteractEvent) event;
            Player player = event1.getPlayer();
            Location playerLoc = player.getLocation();
            Vector direction = playerLoc.getDirection().normalize();

            Location explosionLoc = playerLoc.clone().subtract(direction.clone().multiply(1.5));
            player.getWorld().createExplosion(explosionLoc, 0.0F);

            Location targetLoc = playerLoc.clone().add(direction.clone().multiply(5));
            AbilityUtils.goVelocity(player, targetLoc, 5);
        }
    }

}

File: abilities/src/main/kotlin/io/github/kdy05/physicalFighters/ability/impl/CP9.kt
package io.github.kdy05.physicalFighters.ability.impl

import io.github.kdy05.abilityAPI.ability.AbilityMeta
import io.github.kdy05.abilityAPI.rank.Rank
import io.github.kdy05.abilityAPI.skill.SkillContext
import io.github.kdy05.physicalFighters.ability.PFAbility
import io.github.kdy05.physicalFighters.ability.AbilityUtils
import io.github.kdy05.physicalFighters.util.SoundUtils
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.event.entity.EntityDamageEvent

@AbilityMeta(
    name = "CP9",
    rank = Rank.S,
    guide = [
        "(철괴 공격) 상대에게 고정 대미지를 줍니다.",
        "(철괴 우클릭) 바라보는 방향으로 빠르게 도약합니다.",
        "(패시브) 낙하 대미지를 무시합니다."
    ]
)
class CP9(owner: Player, context: SkillContext) : PFAbility(owner, context) {
    private val activeSkill = object : ActiveSkill() {
        override val cooldownTicks = 400L
        private var pendingAction: (() -> Unit)? = null

        override fun register() {
            onEntityDamage { e ->
                if (owner.inventory.itemInMainHand.type != Material.IRON_INGOT) return@onEntityDamage
                val entity = e.entity as? LivingEntity ?: return@onEntityDamage
                pendingAction = {
                    AbilityUtils.piercingDamage(entity, 6.0)
                    SoundUtils.playBreakSound(entity)
                    owner.sendMessage("${ChatColor.RED}${entity.name}에게 지건을 사용했습니다.")
                }
                activate()
            }
            onRightClick {
                if (owner.inventory.itemInMainHand.type != Material.IRON_INGOT) return@onRightClick
                pendingAction = {
                    val direction = owner.location.direction.normalize()
                    owner.world.createExplosion(owner.location.clone().subtract(direction.clone().multiply(1.5)), 0.0f)
                    AbilityUtils.goVelocity(owner, owner.location.clone().add(direction.clone().multiply(5.0)), 5.0)
                }
                activate()
            }
        }

        override fun onActivate() {
            pendingAction?.invoke()
            pendingAction = null
        }
    }

    private val passiveSkill = object : PassiveSkill() {
        override fun register() {
            on(EntityDamageEvent::class) { e ->
                if (e.entity.uniqueId != owner.uniqueId) return@on
                if (e.cause != EntityDamageEvent.DamageCause.FALL) return@on
                e.isCancelled = true
                SoundUtils.playSuccessSound(owner)
                owner.sendMessage("${ChatColor.GREEN}사뿐하게 떨어져 대미지를 받지 않았습니다.")
            }
        }
    }

    override fun buildSkills() = listOf(activeSkill, passiveSkill)
}
